cpp/dynamic_community_detection_module/algorithm/dynamic_community_detection.cpp

```diff
@@ -0,0 +1,370 @@
+  for (const auto& element : v1) {
+    if (std::find(v2.begin(), v2.end(), element) == v2.end()) return false;
+  }
+
+  return true;
+}
+
+std::unordered_set<std::uint64_t> LabelRankT::InNeighborsMemgraphIDs(
+    std::uint64_t node_id) {
+  std::unordered_set<std::uint64_t> neighbors;
+
+  auto neighbor_set = this->is_directed
+                          ? graph->InNeighbours(graph->GetInnerNodeId(node_id))
+                          : graph->Neighbours(graph->GetInnerNodeId(node_id));
+  for (const auto node_i : neighbor_set) {
+    neighbors.insert(graph->GetMemgraphNodeId(node_i.node_id));
+  }
+
+  return neighbors;
+}
+
+std::vector<std::uint64_t> LabelRankT::NodesMemgraphIDs() {
+  std::vector<std::uint64_t> nodes;
+
+  for (const auto node : graph->Nodes()) {
+    nodes.push_back(graph->GetMemgraphNodeId(node.id));
+  }
+
+  return nodes;
+}
+
+void LabelRankT::SetStructures(std::uint64_t node_id) {
+  times_updated[node_id] = 0;
+  std::unordered_map<std::uint64_t, double> node_label_Ps;
+
+  double sum_w_i = w_selfloop;
+  for (const auto node_j_id : InNeighborsMemgraphIDs(node_id)) {
+    sum_w_i += GetTotalWeightBetween(node_j_id, node_id);
+  }
+
+  // add self-loop
+  node_label_Ps[node_id] = w_selfloop / sum_w_i;
+
+  // add other edges
+  for (const auto node_j_id : InNeighborsMemgraphIDs(node_id)) {
+    node_label_Ps[node_j_id] +=
+        GetTotalWeightBetween(node_j_id, node_id) / sum_w_i;
+  }
+
+  label_Ps[node_id] = node_label_Ps;
+  sum_w[node_id] = sum_w_i;
+}
+
+void LabelRankT::RemoveDeletedNodes(
+    std::unordered_set<std::uint64_t>& deleted_nodes_ids) {
+  for (const auto node_id : deleted_nodes_ids) {
+    label_Ps.erase(node_id);
+    sum_w.erase(node_id);
+    times_updated.erase(node_id);
+  }
+}
+
+double LabelRankT::GetTotalWeightBetween(std::uint64_t from_node_id,
+                                         std::uint64_t to_node_id) {
+  double total_weight = 0;
+  for (auto edge_id :
+       graph->GetEdgesBetweenNodes(graph->GetInnerNodeId(from_node_id),
+                                   graph->GetInnerNodeId(to_node_id))) {
+    auto edge = graph->GetEdge(edge_id);
+
+    // edge direction check
+    if (this->is_directed &&
+        (graph->GetMemgraphNodeId(edge.from) != from_node_id ||
+         graph->GetMemgraphNodeId(edge.to) != to_node_id))
+      continue;
+
+    total_weight += GetWeight(edge_id);
+  }
+
+  return total_weight;
+}
+
+double LabelRankT::GetWeight(std::uint64_t edge_id) {
+  if (!this->is_weighted) return this->DEFAULT_WEIGHT;
+
+  return graph->GetWeight(edge_id);
+}
+
+void LabelRankT::ResetTimesUpdated() {
+  for (auto [node_id, _] : times_updated) {
+    times_updated[node_id] = 0;
+  }
+}
+
+std::int64_t LabelRankT::NodeLabel(std::uint64_t node_id) {
+  std::int64_t node_label = -1;
+  double max_P = 0;
+
+  for (const auto [label, P] : label_Ps[node_id]) {
+    if (P > max_P || (P == max_P && (std::int64_t)label < node_label)) {
```

Comment:
As far as I see, `label` is type of `uint64_t`, thus `(std::int64_t)label < node_label` can only be true if `label > std::numeric_limits<int64_t>::max()`, because it will overflow. Am I missing something here?

Reply:
Casting `label` to `int64_t` didn’t cause any such behavior here. I’ve prepared a pastebin @ https://pastebin.com/kELFfyBt for you to verify – if you run it, `2 2 2` should be printed out.

Reply:
Furthermore, I’ve run the unit tests and verified that `(std::int64_t)label < node_label` is in fact true in some tests.

Reply:
First of all, I missed that `node_label` is overwritten, so it won't be -1 always. Therefore this comment of mine is wrong, sorry for that.

Second, the overflow only happens if the unsigned value is greater than the maximum value of the signed value, [check here](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:2,lang:c%2B%2B,selection:(endColumn:53,endLineNumber:9,positionColumn:53,positionLineNumber:9,selectionStartColumn:53,selectionStartLineNumber:9,startColumn:53,startLineNumber:9),source:'%23include+%3Ciostream%3E%0A%23include+%3Climits%3E%0A+%0Aint+main()+%7B%0A++++std::int64_t+x+%3D+2%3B%0A++++std::uint64_t+y+%3D+std::numeric_limits%3Cstd::uint64_t%3E::max()%3B%0A++++std::int64_t+z+%3D+(std::int64_t)y%3B%0A++++%0A++++std::cout+%3C%3C+x+%3C%3C+!'%5Cn!'+%3C%3C+y+%3C%3C+!'%5Cn!'+%3C%3C+z+%3C%3C+!'%5Cn!'%3B%0A+%0A++++return+0%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%232',t:'0')),k:49.66057441253264,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:clang_trunk,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:2,lang:c%2B%2B,libs:!(),options:'-std%3Dc%2B%2B17+-O1+-DNDEBUG',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:2,tree:'1'),l:'5',n:'0',o:'x86-64+clang+(trunk)+(C%2B%2B,+Editor+%232,+Compiler+%232)',t:'0'),(h:output,i:(compiler:2,editor:2,fontScale:14,fontUsePx:'0',wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+(trunk)+(Compiler+%232)',t:'0')),header:(),k:50.33942558746736,l:'4',m:100,n:'0',o:'',s:1,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).

Reply:
As label numbers are obtained from node IDs (starting at 0), overflow shouldn’t happen unless we’re dealing with a graph with ~2^63 nodes, and hardware is still a long way from there.

I’ll reassess how IDs work anyway so as to alleviate any possible confusion and future-proof the algorithm.